### PR TITLE
Validate config URIs

### LINKED ISSUE #1061: empty string is not a valid Metastore URI
and should therefore be rejected.

Same for `default_index_root_uri`

### PR DESCRIPTION
**Pull request by guilload**
### Description
- Validate config URIs (metastore URI, index URI, index root URI), fixes #1061

### How was this PR tested?
- Added unit tests
- Ran `make test-all`

### CODE REVIEW

File: quickwit-common/Cargo.toml [hunk 1]
```diff
@@ -13,15 +13,17 @@ documentation = "https://quickwit.io/docs/"
 anyhow = "1"
 colored = "2"
 env_logger = "0.9"
+futures = "0.3"
```
Comment (guilload) [inline quickwit-common/Cargo.toml:16]: Sorted the deps and added `serde`.

File: quickwit-config/src/config.rs [hunk 2]
```diff
@@ -387,6 +391,16 @@ impl std::fmt::Debug for QuickwitConfig {
...
+where D: Deserializer<'de> {
+    let uri_opt: Option<String> = Deserialize::deserialize(deserializer)?;
+    uri_opt
+        .map(|uri| Uri::try_new(&uri))
+        .transpose()
+        .map_err(D::Error::custom)
+}
+
 #[cfg(test)]
 mod tests {
```
Comment (guilload) [inline quickwit-config/src/config.rs:405]: Todo: add an extra unit test here for config with invalid URI.

File: quickwit-core/src/index.rs [hunk 3]
```diff
@@ -83,21 +84,23 @@ impl IndexService {
...
             );
-            default_index_uri
+            index_uri
         };
-
         let index_metadata = IndexMetadata {
-            index_id: index_config.index_id.clone(),
-            index_uri,
+            index_id,
+            index_uri: index_uri.into_string(),
```
Comment (guilload) [inline quickwit-core/src/index.rs:103]: `IndexMetadata.index_uri` should be a `Uri` eventually.

File: quickwit-metastore/src/metastore_resolver.rs [hunk 4]
```diff
@@ -126,11 +126,14 @@ impl MetastoreUriResolver {
     }
 
     /// Resolves the given URI.
-    pub async fn resolve(&self, uri: &str) -> Result<Arc<dyn Metastore>, MetastoreResolverError> {
-        let protocol = uri.split("://").next().ok_or_else(|| {
+    pub async fn resolve<S: AsRef<str>>(
```
Comment (guilload) [inline quickwit-metastore/src/metastore_resolver.rs:129]: Adding `S: AsRef<str>` allowed me to not change `.resolve(&metastore_uri)` to `.resolve(metastore_uri.as_ref())` all over the place. We can get rid of it once we change the signature of `resolve(&str)` to `resolve(&Uri)`

File: quickwit-config/src/config.rs (hunk 2 above)
Comment (guilload) [reply quickwit-config/src/config.rs:405]: Done.

File: quickwit-common/src/uri.rs [hunk 5]
```diff
@@ -83,48 +90,94 @@ impl Uri {
...
+                self.uri,
+                path
+            );
+        }
+        let joined = match self.protocol() {
+            FILE_PROTOCOL => Path::new(&self.uri)
+                .join(path)
+                .to_string_lossy()
+                .to_string(),
+            RAM_PROTOCOL | S3_PROTOCOL => format!(
```
Comment (evanxg852000) [inline quickwit-common/src/uri.rs:162]: I would just do this for any other protocol and specially treat `file://` like you have above.
The semantic being that when you join a path, you need it with a separator. we also have `postgres` `postgresql` and many more.

File: quickwit-common/src/uri.rs [hunk 6]
```diff
@@ -83,48 +90,94 @@ impl Uri {
...
+                path
+            ),
+            other => bail!("Protocol `{}` does not support `Uri::join`.", other),
+        };
+        Ok(Self {
+            uri: joined,
+            protocol_idx: self.protocol_idx,
+        })
+    }
+}
```
Comment (evanxg852000) [inline quickwit-common/src/uri.rs:175]: I also found it handy to add 
```rust
pub fn to_str(&self) -> &str {
    self.as_ref()
}
```

File: quickwit-common/src/uri.rs (hunk 5 above)
Comment (guilload) [reply quickwit-common/src/uri.rs:162]: I agree, adding `as_str`.

`join` does not really make sense for `postgres` URIs that is why I wanted those to fall in the catchall category and raise. However, the way I've implemented this function prevents the use of protocols that we have not anticipated, which is bound to break at some point. So I'm going to go with your recommendation for that reason :)

File: quickwit-cli/src/index.rs [hunk 7]
```diff
@@ -513,14 +513,12 @@ pub async fn list_index_cli(args: ListIndexesArgs) -> anyhow::Result<()> {
     debug!(args = ?args, "list");
     let metastore_uri_resolver = quickwit_metastore_uri_resolver();
     let quickwit_config = load_quickwit_config(&args.config_uri, None).await?;
-    let metastore_uri = if let Some(uri) = args.metastore_uri {
-        uri.to_string()
-    } else {
-        quickwit_config.metastore_uri()
-    };
+    let metastore_uri = args
```
Comment (fmassot) [inline quickwit-cli/src/index.rs:516]: Nice, the previous code was "ugly".

File: quickwit-common/Cargo.toml (hunk 1 above)
Comment (fmassot) [reply quickwit-common/Cargo.toml:16]: Is there a tool to sort the deps?

File: quickwit-common/src/uri.rs [hunk 8]
```diff
@@ -83,48 +90,103 @@ impl Uri {
...
         }
     }
+
+    /// Consumes the [`Uri`] struct and returns the normalized URI as a string.
+    pub fn into_string(self) -> String {
+        self.uri
+    }
+
+    /// Creates a new [`Uri`] with `path` adjoined to `self`.
+    /// Fails if `path` is absolute, it replaces the current path.
```
Comment (fmassot) [inline quickwit-common/src/uri.rs:153]: Is `/// Fails if `path` is absolute.` not sufficient for the docs? Do you want to justify why it fails? In this case I would write `Fails if `path` is absolute as we don't want to replace the current path`?

File: quickwit-common/src/uri.rs [hunk 9]
```diff
@@ -83,48 +90,103 @@ impl Uri {
...
         Ok(Self {
             uri: format!("{}{}{}", protocol, PROTOCOL_SEPARATOR, path),
             protocol_idx: protocol.len(),
         })
     }
 
-    /// Returns the URI extension.
+    /// Constructs a [`Uri`] from a properly formatted string `<protocol>://<path>` where `path` is
+    /// normalized. Use this method exclusively for trusted input.
+    pub fn new(uri: String) -> Self {
```
Comment (fmassot) [inline quickwit-common/src/uri.rs:112]: It is strange to make this function public.

File: quickwit-config/src/config.rs [hunk 10]
```diff
@@ -252,6 +255,11 @@ impl QuickwitConfig {
         if self.peer_seeds.is_empty() {
             warn!("Seed list is empty.");
         }
+        let data_dir_uri = Uri::try_new(&self.data_dir_path.to_string_lossy())?;
+
+        if data_dir_uri.protocol() != "file" {
```
Comment (fmassot) [inline quickwit-config/src/config.rs:260]: Use `FILE_PROTOCOL` instead of `"file"`

Comment (guilload) [reply quickwit-config/src/config.rs:260]: Good catch.

File: quickwit-common/src/uri.rs (hunk 8 above)
Comment (guilload) [reply quickwit-common/src/uri.rs:153]: Yes, I don't know what happened here, seems like an unfortunate copy and paste :D
Will fix.

File: quickwit-core/src/index.rs [hunk 11]
```diff
@@ -83,21 +84,23 @@ impl IndexService {
...
-        let index_uri = if let Some(index_uri) = index_config.index_uri.as_ref() {
-            index_uri.to_string()
+        let index_id = index_config.index_id.clone();
+        let index_uri = if let Some(index_uri) = &index_config.index_uri {
+            index_uri.clone()
         } else {
-            let default_index_uri =
-                format!("{}/{}", self.default_index_root_uri, index_config.index_id);
+            let index_uri = self.default_index_root_uri.join(&index_id).expect(
+                "Failed to create default index URI. This should never happen. Please, report.",
```
Comment (fmassot) [inline quickwit-core/src/index.rs:92]: Don't know if it's a good idea to put the github issue link.

File: quickwit-common/src/uri.rs (hunk 9 above)
Comment (guilload) [reply quickwit-common/src/uri.rs:112]: Yes. I don't like that it opens the door for shooting yourself in the foot. At the same time, sometimes you know the input is trustworthy (deserialization from metastore), and that's really what you want to use. It's also very convenient for writing less verbose unit tests.

File: quickwit-core/src/index.rs (hunk 11 above)
Comment (guilload) [reply quickwit-core/src/index.rs:92]: Might as well.

File: quickwit-core/src/index.rs (hunk 3 above)
Comment (fmassot) [reply quickwit-core/src/index.rs:103]: Indeed, we could open an issue to track that.

File: quickwit-common/src/uri.rs (hunk 9 above)
Comment (fmassot) [reply quickwit-common/src/uri.rs:112]: ok, so at least there is one useful use case. 
For unit tests, we can make it available only for that.

**Review by fmassot — APPROVED** [review]
Nice job as always, log messages are way better than before.

File: quickwit-common/src/uri.rs (hunk 9 above)
Comment (guilload) [reply quickwit-common/src/uri.rs:112]: I don't know if we can do it across crates of the same workspace. Last time I tried I failed.

File: quickwit-common/Cargo.toml (hunk 1 above)
Comment (guilload) [reply quickwit-common/Cargo.toml:16]: The cargo team is still working on it.

**Pull request merged**